### ISSUE TITLE
The strided matmul template mixes a scalar pid3 with a uint3 _lid3 on a flat grid

### ISSUE THREAD

**Issue by benkelaya** [bledden/triton-msl#10]
**Author:** Hocine Benkelaya (with NeuroBrix)

## The point that decides

`_lowerer_templates.py`, in `_lower_strided_scalar_matmul`:

```python
if has_pid:
    self._used_pid_axes = {0, 1}
    # Metal requires all thread-index attributes to share a type; use
    # uint3 for both when multi-axis dispatch is in play.
    _flat_grid = self._emit_tile_ids_from_source_grid(lines, None)
    lines.append("    uint3 _lid3 [[thread_position_in_threadgroup]]")
```

The comment states the invariant exactly. `_lid3` is then emitted as `uint3`
**unconditionally**, while `pid3`'s arity is decided independently a few
hundred lines earlier, in `_emit_tile_ids_from_source_grid`:

```python
self._pid3_is_vector = bool(_batch_axes) or len(self._tile_axes) > 1
...
if self._pid3_is_vector:
    lines.append("    uint3 pid3 [[threadgroup_position_in_grid]],")
else:
    lines.append("    uint pid3 [[threadgroup_position_in_grid]],")
```

So the invariant holds only when the source grid happens to be multi-axis. On a
**flat** grid — one tile axis, no batch — `pid3` is scalar, `_lid3` stays
`uint3`, and the pair is exactly what the comment says Metal forbids.

## What is actually emitted

```
error: expecting input declarations with either all scalar types or all vector
       types with the same number of elements
kernel void matmul_kernel(
            ^
note: declaration with attribute 'threadgroup_position_in_grid' of type 'uint'
    uint pid3 [[threadgroup_position_in_grid]],
note: declaration with attribute 'thread_position_in_threadgroup' of type
      'uint3' (vector of 3 'unsigned int' values)
    uint3 _lid3 [[thread_position_in_threadgroup]]
1 error generated.
```

The shader does not build. The kernel then falls back to CPU with a warning, so
a caller sees a run that is correct and very slow rather than an error — which
is why it can sit unnoticed.

## Reproducer

A single-dot matmul with a **non-contiguous B** (what routes to this template)
and a **flat grid** (what makes `pid3` scalar).

```python
import torch, triton, triton.language as tl

@triton.jit
def mm(a_ptr, b_ptr, c_ptr, M, N, K,
       sam, sak, sbk, sbn, scm, scn,
       BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr):
    pid = tl.program_id(0)                 # ONE tile axis: a flat grid
    n_tiles = (N + BLOCK_N - 1) // BLOCK_N
    pm, pn = pid // n_tiles, pid % n_tiles
    offs_m = pm * BLOCK_M + tl.arange(0, BLOCK_M)
    offs_n = pn * BLOCK_N + tl.arange(0, BLOCK_N)
    offs_k = tl.arange(0, BLOCK_K)
    acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=tl.float32)
    for k in range(0, K, BLOCK_K):
        a = tl.load(a_ptr + offs_m[:, None] * sam + (k + offs_k)[None, :] * sak)
        b = tl.load(b_ptr + (k + offs_k)[:, None] * sbk + offs_n[None, :] * sbn)
        acc += tl.dot(a, b)
    tl.store(c_ptr + offs_m[:, None] * scm + offs_n[None, :] * scn, acc)

M = N = K = 64
BM = BN = BK = 32
a = torch.randn(M, K, device="mps")
b = torch.randn(N, K, device="mps").t()    # inner dim NOT contiguous
c = torch.zeros(M, N, device="mps")
mm[((M // BM) * (N // BN),)](
    a, b, c, M, N, K,
    a.stride(0), a.stride(1), b.stride(0), b.stride(1), c.stride(0), c.stride(1),
    BLOCK_M=BM, BLOCK_N=BN, BLOCK_K=BK)
```

Making `b` contiguous, or giving the kernel two tile axes, both make it build.

## Why this surfaced now

Nothing in the template changed. A caller started walking a **pre-transposed
weight by its strides** — `x @ w.t()` without materialising the transpose,
which is precisely the case `_matmul_stride_decision`'s docstring names as the
reason this template exists:

> *at least one operand has a NON-contiguous inner dim (col stride is a runtime
> arg or a non-unit constant, e.g. a transposed B from `x @ w.t()`)*

So the route is being used as designed, and the signature it emits is
ill-formed whenever the grid is flat. The two conditions are independent, which
is why this can live a long time: a strided operand with a 2-D grid builds.

## Suggested shape, not a patch

The arity is a property of the *signature*, not of either half, so it wants
deciding once. `_emit_tile_ids_from_source_grid` already computes
`_pid3_is_vector`; `_lid3` could read it, or — since the comment's rule is
"share a type" and `_lid3.x` is what the body uses anyway — `pid3` could be
emitted `uint3` on this path unconditionally, matching what the comment says
was intended. Which of the two is right depends on what else reads
`_pid3_is_vector`, and that is yours rather than mine.

## Environment

Apple M4 Pro, macOS 26.6.2 (Darwin 25.6.0), Python 3.12.0, triton 3.7.0,
Metal backend.

**Comment by bledden** [comment]
Thanks for tracing this. The signature invariant you identified is real, but the named helpers that mixed the scalar and vector types belonged to the unmerged PR #6 branch, rather than the released backend. The released path uses matching `uint3` declarations, so I didn't port a repair for those fork-only helpers.

Separately, the plain-flat-grid gap exposed by your non-contiguous-B example is addressed by 0.3.0's proven mapping path. Strided operands and flat grids need to be checked together; making B contiguous isn't a requirement I'm asking you to accept as the fix. I also haven't treated your reported CPU fallback as proof that the Metal path succeeded. My earlier probe of the released lineage refused explicitly rather than emitting that malformed signature or silently executing it on CPU.

If the shader error or unexpected fallback remains on 0.3.0, please include the fork commit and warning as well as the package version so I can distinguish the two implementations and the actual execution route.

**Comment by benkelaya** [comment]
Thanks for pointing out that those helpers only lived on our #6 branch, understood. If the shader error or a fallback shows up again, we'll include the package version, fork commit and warning.
